metricq_tools/summary.py

```diff
@@ -0,0 +1,294 @@
+from metricq.subscriber import Subscriber
+from tabulate import tabulate
+
+from metricq_tools.utils import metricq_server_option, metricq_token_option
+
+from .logging import get_root_logger
+from .version import version as client_version
+
+logger = get_root_logger()
+
+click_completion.init()
+
+
+class Summary:
+    def __init__(
+        self,
+        metrics: list[str],
+        intervals_histogram: bool,
+        values_histogram: bool,
+        print_stats: bool,
+        print_data: bool,
+    ):
+        self._metrics = metrics
+
+        self.print_intervals = intervals_histogram
+        self.print_values = values_histogram
+        self.print_stats = print_stats
+        self.print_data = print_data
+
+        self.timestamps = dict[str, list]()
+        self.last_timestamp = dict[str]()
+        self.intervals = dict[str, list]()
+        self.values = dict[str, list]()
+
+        for metric in metrics:
+            self.timestamps[metric] = list()
+            self.last_timestamp[metric] = None
+            self.intervals[metric] = list()
+            self.values[metric] = list()
+
+    async def add_data(self, metric: str, timestamp: metricq.Timestamp, value: float):
+        if self.print_data:
+            click.echo(click.style("{}: {}".format(timestamp, value), fg="bright_blue"))
+
+        self.timestamps[metric].append(timestamp.posix)
+        if self.last_timestamp[metric]:
+            self.intervals[metric].append(timestamp.posix - self.last_timestamp[metric])
+        self.last_timestamp[metric] = timestamp.posix
+        self.values[metric].append(value)
+
+    def _print_histogram(self, values):
+        counts, bin_edges = np.histogram(values, bins="doane")
+        fig = tpl.figure()
+        labels = [
+            "[{:#.6g} - {:#.6g})".format(bin_edges[k], bin_edges[k + 1])
+            for k in range(len(bin_edges) - 2)
+        ]
+        labels.append(
+            "[{:#.6g} - {:#.6g}]".format(
+                bin_edges[len(bin_edges) - 2], bin_edges[len(bin_edges) - 1]
+            )
+        )
+        fig.barh(counts, labels=labels)
+        fig.show()
+
+    def _print_intervals_histogram(self, intervals):
+        click.echo(
+            click.style(
+                "Distribution of the duration between consecutive data points in seconds",
+                fg="yellow",
+            )
+        )
+        click.echo()
+
+        self._print_histogram(intervals)
+
+        click.echo()
+        click.echo()
+
+    def _print_values_histogram(self, values):
+        click.echo(
+            click.style("Distribution of the values of the data points", fg="yellow")
+        )
+        click.echo()
+
+        self._print_histogram(values)
+
+        click.echo()
+        click.echo()
+
+    def print(self):
+        for metric in self._metrics:
+
+            if self.values[metric]:
+                click.echo()
+                click.echo(click.style(f"Stats of metric {metric!r}:", fg="green"))
```

Comment:
This header is redundant if nothing gets printer afterwards. It should not be printed if none of the conditions below are true.

Please also spell out *Statistics*, or even do `f"Statistics for {metric}"`.